### PR TITLE
Add Qualcomm Hexagon QNN HTP backend

### LINKED ISSUE #77: Build the Qualcomm Hexagon NPU backend
## Goal

Add a production-oriented Qualcomm Hexagon NPU backend that implements `virtio_accel_core::Accelerator` and executes device-neutral TOSA 1.0 programs on supported Qualcomm NPUs without leaking Qualcomm runtime types into the portable crates.

## Scope

- Add a separate host-native crate (provisionally `virtio-accel-hexagon`) for Qualcomm Hexagon integration.
- Select and document the supported public Qualcomm userspace runtime/API, required device families, required driver/runtime versions, and version policy.
- Probe the native runtime at build time and provide a compile-only placeholder when it is unavailable, following the portability pattern used by existing host backends.
- Discover Qualcomm NPU devices and report an honest NPU identity, capabilities, memory domains, alignments, and resource limits.
- Accept `virtio-accel-tosa` artifacts, validate/analyze them with the shared TOSA layer, and lower a clearly documented initial operator/dtype subset to the Qualcomm runtime executable representation.
- Implement contexts, provider-owned buffers, programs, execution queues, submissions, events, transfers, release, and recovery through the shared `Accelerator` contract.
- Bind the exact provider allocation for program-visible inputs, outputs, and mutable state. Do not introduce submission-time bounce buffers; explicit `read_buffer`/`write_buffer` remain the copy boundaries.
- Preserve rejected versus indeterminate submission/release semantics, stable nonblocking event polling, timeout behavior, device-loss reporting, and cancellation only if the runtime can implement it honestly.
- Confine unsafe FFI code to an audited native boundary and add a `SAFETY.md` covering handle ownership, callbacks/concurrency, buffer lifetimes, and teardown.
- Add a backend-local end-to-end example and update the workspace, support matrix, portability guide, architecture notes, performance evidence, and release policy.

## Existing backend parity references

Use the current backend model and behavior as the parity baseline:

- Apple Core ML / ANE (`virtio-accel-coreml`)
- Intel OpenVINO (`virtio-accel-openvino`)
- Mock backend infrastructure (`virtio-accel-mock`, deterministic harness behavior only)

## Initial support boundary

The first hardware-executable tier should cover the shared static TOSA FP32/FP16 identity, non-square batched MATMUL, and NHWC MAX_POOL2D cases where the Qualcomm compiler/runtime supports them. Any operator, dtype, profile, extension, dynamic shape, or layout outside the advertised subset must be rejected during program admission.

Do not silently:

- fall back to CPU or GPU execution;
- reduce declared precision;
- dequantize unsupported FP8, INT8, or packed INT4 graphs; or
- stage program-visible buffers during submission.

If an initial corpus case is not supported by the target hardware/runtime, document the limitation explicitly and keep the backend's advertised target narrower rather than skipping it silently.

## Acceptance criteria

- The entire workspace builds and tests on hosts without Qualcomm dependencies; the native path is enabled only when its requirements are detected or explicitly requested.
- On a supported Qualcomm NPU host, the backend enumerates a Qualcomm Hexagon device as `AcceleratorClass::NPU` and completes the documented TOSA example end to end.
- The backend passes every mandatory case in `virtio-accel-conformance`; capability-conditional skips include explicit reasons.
- Every advertised operator/dtype case from the shared numerical TOSA corpus executes on the NPU and matches its oracle.
- Tests prove exact direct binding of caller-owned provider allocations and report transfer/staging diagnostics through the conformance hook.
- Tests cover malformed and unsupported artifact rejection, incompatible bindings, concurrent binding conflicts, finite timeouts, stable terminal polling, device loss, and rejected/indeterminate ownership paths that the runtime can produce.
- Native resources are released exactly once across normal teardown, event completion, failure, and backend discard; diagnostics return to baseline after each conformance case.
- CI includes portable compile/lint/unit coverage. A hardware lane is added when a Qualcomm NPU runner is available; until then, the required driver/runtime versions and reproducible hardware test commands are documented.
- The README support table changes Qualcomm Hexagon from `Planned` only for the capabilities demonstrated by tests.

## Non-goals

- Changes to the virtio wire ABI or portable `Accelerator` contract solely to mirror a Qualcomm API.
- External-memory/DMA-BUF import, cross-process sharing, or new synchronization primitives.
- Full TOSA operator coverage or low-precision support beyond what is explicitly implemented and tested.
- CPU/GPU fallback inside the Qualcomm NPU backend.

## Implementation guidance (backend-agnostic, Qualcomm-focused)

Because Qualcomm software stacks vary by platform, start by selecting one documented runtime path and proving one strict vertical slice end to end before adding variants.

### Recommended first vertical slice

1. Parse and validate one static TOSA graph with `virtio-accel-tosa`.
2. Match it to a small, explicit lowering subset with a fixed binding order.
3. Compile/prepare executable artifacts during `load_program`, never during `submit`.
4. Create runtime executable state once per admitted program.
5. Allocate persistent provider-owned buffers and pass those exact allocations to dispatch.
6. Execute through a serialized worker lane and expose completion through nonblocking `poll_event`.

### Map backend objects directly

Keep a one-to-one mapping that preserves ownership and lifetime clarity:

| `Accelerator` object | Qualcomm backend representation |
| --- | --- |
| Backend instance | One detected Qualcomm NPU device plus a serialized native execution lane |
| Context | Lightweight ownership/accounting scope without unnecessary process-global duplication |
| Buffer | One native allocation with persistent mapping and exact retained bytes/alignment |
| Program | Validated TOSA metadata, slot/shape plan, compiler/runtime identity, and executable handle |
| Queue | Bounded admission queue feeding the serialized native dispatch lane |
| Event | Atomic/latched state plus owned references to the program and all bound native buffers until terminal release |

For explicit writes, copy into the mapped native allocation and perform any runtime-required synchronization before dispatch. For explicit reads, perform runtime-required invalidation/synchronization before copying out. Submission must use the same caller-bound provider allocations and exact bound ranges.

## Follow-up work (out of initial slice)

- Additional device families and runtime adapters once the first path is stable and contract-compliant.
- Optional low-precision tiers (INT8/INT4/FP8) only when direct execution is demonstrated and conformance-backed.
- Performance tuning after correctness, ownership, and conformance invariants are locked.

Related: #75

### PR DESCRIPTION
**Pull request by Colin-Hayes**
## Summary

- add the `virtio-accel-hexagon` Windows ARM64 host backend using the public QAIRT/QNN HTP interface
- validate static FP16 TOSA and execute identity, batched MATMUL, and NHWC MAX_POOL2D without CPU/GPU fallback
- bind exact aligned caller allocations, serialize native execution, and expose stable nonblocking events
- preserve an explicit SDK-free placeholder on unsupported hosts
- add an audited native boundary, SDK/manual-test documentation, and release-policy integration
- add a reusable two-sample FP16 linear-classifier fixture and runnable Hexagon example

## Hardware evidence

Validated on Snapdragon X126100 / Hexagon HTP v73 with:

- QAIRT `2.49.0.260730`
- provider `HTP_QTI_AISW`
- provider build `v2.49.0.260730134355`
- QNN core API `2.38.0`
- HTP backend API `5.49.0`

The five hardware integration tests pass, including the shared numerical corpus and mock classifier. The classifier produces the expected FP16 logits `[4.0, -1.0, 1.0, -1.5]`.

## Verification

- `cargo fmt --all -- --check`
- `cargo clippy --workspace --all-targets --all-features -- -D warnings` with the Hexagon backend forced off
- `cargo test --workspace --all-targets --all-features` with the Hexagon backend forced off
- `cargo clippy -p virtio-accel-hexagon --all-targets -- -D warnings` against QAIRT
- `cargo test -p virtio-accel-hexagon --all-targets -- --test-threads=1` against physical HTP
- `cargo run -p virtio-accel-hexagon --example mock_classifier` against physical HTP
- `python ci/check-release-policy.py`
- `cargo package -p virtio-accel-hexagon --allow-dirty`

## Current boundary

The advertised tier is static FP16 only. Finite timeouts are rejected before admission and cancellation is not advertised. Direct PoolMax2d finalization is rejected by this Windows HTP runtime, so its zero-padding semantics are lowered to HTP Gather plus ElementWiseMaximum nodes. Full conformance hooks, formal direct-binding diagnostics, device-loss injection, stress accounting, performance evidence, and a controlled hardware CI lane remain tracked in the implementation plan.

Relates to #77

### CODE REVIEW

File: crates/virtio-accel-hexagon/src/native.rs [hunk 1]
```diff
@@ -0,0 +1,962 @@
...
+                        FeatureRole::Output => ffi::TENSOR_OUTPUT,
+                    });
+                ffi::TensorDesc {
+                    value: tensor.value,
+                    role,
+                    dimensions: tensor.dims.as_ptr(),
+                    rank: tensor.dims.len() as u32,
+                }
+            })
+            .collect::<Vec<_>>();
```
[the remark this answers is not in the document]
Comment (Colin-Hayes) [reply crates/virtio-accel-hexagon/src/native.rs:734]: Fixed in 09e38b9. The Rust/C ABI now carries the explicit model I/O index, and the QNN bridge builds its binding tables by that index while rejecting duplicates and gaps. Added a regression test with tensor declaration order differing from model I/O order; the hardware-backed HTP suite passes.

File: crates/virtio-accel-hexagon/build.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        return;
+    }
+
+    println!("cargo:rerun-if-changed={}", header.display());
+    println!("cargo:rerun-if-changed=native/qnn_bridge.cpp");
+    println!("cargo:rerun-if-changed=native/qnn_bridge.h");
+    cc::Build::new()
+        .cpp(true)
+        .file("native/qnn_bridge.cpp")
+        .include(root.join("include/QNN"))
```
[the remark this answers is not in the document]
Comment (Colin-Hayes) [reply crates/virtio-accel-hexagon/build.rs]: Fixed in 09e38b9. SDK detection now selects the include directory only when the complete QnnInterface/QnnOpDef/HTP header set is present, and passes that detected directory to the C++ compiler. Both include/QNN and flat include layouts compile successfully.

**Review by SnowCheetos — APPROVED** [review]
